src/hashes/ternary/curl_p/batched/mod.rs

```diff
@@ -0,0 +1,219 @@
+    /// binary encoding of the output hashes. Each pair of low and high bits in the `bct_hashes`
+    /// field is decoded into a trit using the same convention as the `mux` step with an additional
+    /// rule for the `(0, 0)` pair of bits which is mapped to the `0` trit.
+    fn demux(&mut self, index: usize) -> TritBuf {
+        for (bc_trit, btrit) in self.bct_hashes.iter().zip(self.buf_demux.iter_mut()) {
+            let lo = (bc_trit.lo() >> index) & 1;
+            let hi = (bc_trit.hi() >> index) & 1;
+
+            *btrit = match (lo, hi) {
+                (1, 0) => Btrit::NegOne,
+                (0, 1) => Btrit::PlusOne,
+                // This can only be `(0, 0)` or `(1, 1)`.
+                _ => Btrit::Zero,
+            };
+        }
+
+        self.buf_demux.clone()
+    }
+
+    /// Hash the received inputs using the batched version of CurlP.
+    ///
+    /// This function also takes care of cleaning the buffers of the struct and resetting the
+    /// batched CurlP hasher so it can be called at any time.
+    pub fn hash_batched(&mut self) -> impl Iterator<Item = TritBuf> + '_ {
+        let total = self.trit_inputs.len();
+        // Reset batched CurlP hasher.
+        self.bct_curlp.reset();
+        // Multiplex the trits in `trits` and dump them into `inputs`
+        self.mux();
+        // Do the regular sponge steps.
+        self.bct_curlp.absorb(&self.bct_inputs);
+        self.bct_curlp.squeeze_into(&mut self.bct_hashes);
+        // Clear the `trits` buffer to allow receiving a new batch.
+        self.trit_inputs.clear();
+        // Fill the `inputs` buffer with zeros.
+        self.bct_inputs.fill(0);
+        // Return an iterator for the output hashes.
+        BatchedHashes {
+            hasher: self,
+            range: 0..total,
+        }
+    }
+
+    /// Hash the received inputs using the regular version of CurlP.
+    ///
+    /// In particular this function does not use the `bct_inputs` and `bct_hashes` buffers, takes
+    /// care of cleaning the `trit_inputs` buffer and resets the regular CurlP hasher so it can be
+    /// called at any time.
+    pub fn hash_unbatched(&mut self) -> impl Iterator<Item = TritBuf> + '_ {
+        self.curlp.reset();
+        UnbatchedHashes {
+            curl: &mut self.curlp,
+            trit_inputs: self.trit_inputs.drain(..),
+        }
+    }
+}
+
+/// A helper iterator type for the output of the `hash_batched` method.
+struct BatchedHashes<'a, B: RawEncodingBuf> {
+    hasher: &'a mut BatchHasher<B>,
+    range: std::ops::Range<usize>,
+}
+
+impl<'a, B> Iterator for BatchedHashes<'a, B>
+where
+    B: RawEncodingBuf,
+    B::Slice: RawEncoding<Trit = Btrit>,
+{
+    type Item = TritBuf;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let index = self.range.next()?;
+        Some(self.hasher.demux(index))
+    }
+}
+
+/// A helper iterator type for the output of the `hash_unbatched` method.
+struct UnbatchedHashes<'a, B: RawEncodingBuf> {
```

Comment:
Do we still need this?

Reply:
I think I would keep it, there is still a threshold (of number of provided hashes) where it's more efficient to not batch

Reply:
It was made like that for situations where you don't know in advance how many things you're going to hash, tbh I don't know if this is still a desired property or not

Reply:
I guess we can keep it for now and get rid of it once it becomes a burden to maintain.

Reply:
I wouldn't create breaking changes in the future if not absolutely necessary. Maybe you are right, the only use case for this now is PoW and we always have 64 inputs ready for it, maybe we should remove